docs/proposals/11-apiserver-network-proxy.md

```diff
@@ -0,0 +1,134 @@
+
+We want to replace this:
+
+`APIServer | VPN-seed ---> internet ---> LB --> VPN-Shoot (4314) --> Pods | Nodes | Services`
+
+With this:
+
+`APIServer <-> Proxy-Server <--- internet <--- Proxy-Agent --> Pods | Nodes | Services`
+
+
+### API Server Network Proxy
+
+To solve this issue we can utilize the [apiserver-network-proxy](https://github.com/kubernetes-sigs/apiserver-network-proxy) upstream implementation. Which provides a reference implementation for a reverse streaming server. The way it works is as follows:
+
+- Proxy agent connects to proxy server to establish a sticky connection.
+- Traffic to the proxy server (residing in the seed) gets then re-directed to the agent (residing in the shoot) which forwards the traffic to in-cluster components.
+
+The initial motivation for the apiserver-network-proxy project is to get rid of provider-specific implementations that reside in the API-server (e.g., SSH), but it turns out that
+it has other interesting use-cases such as data-plane connection decoupling, which is the main use-case for this proposal.
+
+Starting with **Kubernetes 1.18** it's possible to make use of an `--egress-selector-config-file` flag, this helps point the API-server to traffic hook points based on traffic direction. For example, in the config below the API server would have to forward all cluster related traffic (e.g., logs, port-forward, exec, ...etc.) to the **proxy-server** which then knows how to forward traffic to the shoot. For the rest of the traffic, e.g. API server to ETCD or other control-plane components `direct` is used which means legacy routing method, i.e., by-pass the proxy.
+```yaml
+  egress-selector-configuration.yaml: |-
+    apiVersion: apiserver.k8s.io/v1alpha1
+    kind: EgressSelectorConfiguration
+    egressSelections:
+    - name: cluster
+      connection:
+        proxyProtocol: httpConnect
+        transport:
+          tcp:
+            url: https://proxy-server:8131
+    - name: master
+      connection:
+        proxyProtocol: direct
+    - name: etcd
+      connection:
+        proxyProtocol: direct
+```
+
+## Challenges
+
+
+### Prometheus to Shoot connectivity
+One challenge remains to completely eliminate the need for a VPN connection. In today's Gardener setup, each control-plane has a Prometheus instance that directly scrapes cluster components such as CoreDNS, Kubelets, cadvisor, etc. This works because in addition to the VPN side car attached to the API server pod, we have another one attached to prometheus which knows how to forward traffic to these endpoints. Once the VPN is eliminated, it is required to find other means to forward traffic to these components.
+
+#### Possible Solutions
+
+There are currently two ways to solve this problem:
+
+- Attach a port-forwarder side-car to prometheus.
+- Utilize the proxy subresource on the API server.
+
+#### Port-forwarder Sidecar
```

Comment:
I would remove this entire section. The more sane option would be to deploy a very simple sidecar proxy to the Prometheus which is going to have a Dialer which would simply do a HTTP CONNECT call to the Proxy-Server, similar to what [K8S's API server is doing](https://github.com/kubernetes/kubernetes/blob/bac9351c64671ce4d5198d431c97bf1ccd72752f/staging/src/k8s.io/apiserver/pkg/server/egressselector/egress_selector.go#L104-L129). This would establish the TCP connection which can then be used to send the actual data of the request.

The setup would be:

```
Prom Pod (Prometheus -> Proxy) <-> Proxy-Server <--- internet <--- Proxy-Agent --> Pods | Nodes | Services
```

Reply:
yup thats a third option, which I actually trying doing but its more code overhead (i.e., another piece to maintain, extra certificates if the client is a side-car and not baked in). I can add it as a third option but I don't think that will be our first go-to approach